File: java/aliennumbers.java
import java.util.Scanner;

public class aliennumbers {
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		int T = scan.nextInt();
		for (int i = 1; i <= T; i++) {
			String alien_number = scan.next();
			String source_lang = scan.next();
			String target_lang = scan.next();
			int toBase10 = 0;
			for (int j = 0; j < alien_number.length(); j++)
				toBase10 += source_lang.indexOf(alien_number.charAt(j))
						* Math.pow(source_lang.length(), alien_number.length() - j - 1);
			StringBuilder translated_number = new StringBuilder();
			while (toBase10 > 0) {
				int d = toBase10 % target_lang.length();
				translated_number.insert(0, target_lang.charAt(d));
				toBase10 /= target_lang.length();
			}
			System.out.printf("Case #%d: %s%n", i, translated_number);
		}
		scan.close();
	}
}

File: ts/src/aliennumbers.ts
import readline = require('readline')

function translate (
  alienNum: string,
  sourceLang: string,
  targetLang: string
): string {
  return toTarget(toBase10(alienNum, sourceLang), targetLang)
}

function toBase10 (alienNum: string, sourceLang: string): number {
  return [...alienNum]
    .map((n) => sourceLang.indexOf(n))
    .reverse()
    .reduce((acc, curr, index) => {
      return acc + curr * Math.pow(sourceLang.length, index)
    })
}

function toTarget (num: number, targetLang: string): string {
  const str: string[] = []
  while (num > 0) {
    str.push(targetLang[num % targetLang.length] ?? '')
    num = Math.floor(num / targetLang.length)
  }
  return str.reverse().join('')
}

let lineCount = 0

readline
  .createInterface({
    input: process.stdin,
    output: process.stdout
  })
  .on('line', (line: string) => {
    if (lineCount > 0) {
      const [alienNum, sourceLang, targetLang] = line.split(' ')
      console.log(
        `Case #${lineCount}: ${translate(
          alienNum ?? '',
          sourceLang ?? '',
          targetLang ?? ''
        )}`
      )
    }
    lineCount++
  })
  .on('close', () => {
    process.exit(0)
  })
